Revit_Core_Engine/Query/UniqueParameterNames.cs

```diff
@@ -0,0 +1,179 @@
+
+                foreach (var elementsByCat in elementsByCategoryDictionary)
+                {
+                    List<Element> elementsOfCat = elementsByCat.Value;
+
+                    if (!elementsOfCat.Any()) 
+                        continue;
+
+                    HashSet<int> instanceParameterIds = elementsOfCat.UniqueParametersIds();
+                    IEnumerable<string> instanceParameterNames = instanceParameterIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in instanceParameterNames)
+                        parameterNames.Item1.Add(name);
+
+                    IEnumerable<Element> elementTypes = elementsOfCat.UniqueTypeIds().Select(x => doc.GetElement(new ElementId(x)));
+                    HashSet<int> typeParametereIds = elementTypes.UniqueParametersIds();
+                    IEnumerable<string> typeParameterNames = typeParametereIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in typeParameterNames)
+                        parameterNames.Item2.Add(name);
+                }
+            }
+
+            return parameterNames;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per document.")]
+        private static Dictionary<Document, List<Element>> ElementsByDocumentDictionary(this IEnumerable<Element> elements)
+        {
+            Dictionary<Document, List<Element>> elementsByDoc = new Dictionary<Document, List<Element>>();
+
+            foreach (Element element in elements)
+            {
+                if (elementsByDoc.ContainsKey(element.Document))
+                    elementsByDoc[element.Document].Add(element);
+                else
+                    elementsByDoc.Add(element.Document, new List<Element> { element });
+            }
+
+            return elementsByDoc;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per category.")]
+        private static Dictionary<ElementId, List<Element>> ElementsByCategoryDictionary(this IEnumerable<Element> elements)
+        {
+            Dictionary<ElementId, List<Element>> elementsByCat = new Dictionary<ElementId, List<Element>>();
+
+            foreach (Element element in elements)
+            {
+                if (elementsByCat.ContainsKey(element.Category.Id))
+                    elementsByCat[element.Category.Id].Add(element);
+                else
+                    elementsByCat.Add(element.Category.Id, new List<Element> { element });
+            }
+
+            return elementsByCat;
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique parameter ids for the collection of the elements.")]
+        private static HashSet<int> UniqueParametersIds(this IEnumerable<Element> elementsFromOneDocument)
+        {
```

Comment:
Here you could have one more hashset that would capture visited families to check element types only once per family:

```C#
HashSet<int> ids = new HashSet<int>();
HashSet<string> visitedFamilies = new HashSet<string>();
foreach (Element element in elementsFromOneDocument)
{
    
    if (element is ElementType)
    {
        string familyName = ((ElementType).element).FamilyName;
        if (visitedFamilies.Contains(familyName)
            continue;
        else
            visitedFamilies.Add(familyName);
    }
    
    foreach (Parameter parameter in element.ParametersMap)
    {
        ids.Add(parameter.Id.IntegerValue);
    }
}
```

**IMPORTANT:** highlight it only works for elements of same category!!

Reply:
Added 👍